rfcs/0004-fulltext-search.md

```diff
@@ -0,0 +1,275 @@
+  and location for the language-specific stems found in the text documents 
+  being indexed. 
+- _ranking algorithm_: "Ranking attempts to measure how relevant documents 
+  are to a particular query, so that when there are many matches the most 
+  relevant ones can be shown first." [- Postgres Documentation](https://www.postgresql.org/docs/11/textsearch-controls.html#TEXTSEARCH-RANKING-SEARCH-RESULTS)
+  
+  **Algorithms**:
+  - _standard ranking_: ranking based on the number of matching lexemes.     
+  - _cover density ranking_: Cover density is similar to the standard 
+    fulltext search ranking except that the proximity of matching lexemes 
+    to each other is taken into consideration. This function requires 
+    lexeme positional information to perform its calculation, so it ignores 
+    any "stripped" lexemes in the index.            
+
+## Detailed Design
+
+### Subgraph Definition
+
+Part of the power of fulltext search indexes are their flexibility, so 
+it is important to expose a simple interface to facilitate useful 
+applications of the index and ulimately aim to reduce the need to create 
+new subgraphs for the express purpose of updating fulltext search fields 
+for end-use cases. 
+
+For each fulltext search index a subgraph developer must be able to specify:
+    1. a language, 
+    2. a set of text document fields to include,
+    3. relative weighting for each field in the index.
+    4. a choice of ranking algorithm for sorting query result items
+
+The proposed process of adding one or more fulltext search index to an Entity 
+involves adding one or more fulltext directive to the `_Schema_` type in the 
+subgraph's GraphQL schema.  Each fulltext definition will have four required
+top level parameters: `name`, `language`, `algorithm`, and `include`. 
+The fulltext search definitions will be used to generate queries on the GraphQL
+schema that will be exposed to the end user.  
+
+Applying fulltext search indexes across entities will be a powerful 
+abstraction allowing users to search across all relevant entities in one
+query. An index search across entities will by definition have polymorphic
+results, so in that case a union type will also be generated in the schema 
+for the query result items. 
+
+Verification of the index definition will ensure that all fields referenced 
+are valid String type fields. To be clear, the proposed interface will add 
+the fulltext search index definitions to the subgraph deployment hash, so 
+adding or updating a fulltext search index will require an update to the subgraph. 
+With subgraph composition it will be possible to easily create new subgraphs 
+that add specific fulltext search indexes to an existing subgraph. 
+
+Example fulltext search definition:
+
+```graphql
+type _Schema_ 
+  @fulltext(
+    name: "media"
+    ...
+  )
+  @fulltext(
+    name: "search",
+    language: "english",
+    algorithm: "ranked",
```

Comment:
If we generate a `_FulltextAlgorithm` or similar enum type, we can use `ranked` instead of `"ranked"` here:
```
enum _FulltextAlgorithm {
  ranked
}
```